src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm

```diff
@@ -635,64 +635,76 @@ bool GetPeerNodeInfo(CommandHandler * commandHandler, const ConcreteCommandPath
+
+                // If the protocol requested is not supported, return status - Protocol Not Supported
+                if (!isBDXProtocolSupported) {
+                    Commands::QueryImageResponse::Type response;
+                    response.status
+                        = static_cast<OTAQueryStatus>(MTROTASoftwareUpdateProviderOTAQueryStatusDownloadProtocolNotSupported);
+                    handler->AddResponse(cachedCommandPath, response);
+                    handle.Release();
+                    return;
+                }
 
-                    char uriBuffer[kMaxBDXURILen];
-                    MutableCharSpan uri(uriBuffer);
-                    err = bdx::MakeURI(targetNodeId.GetNodeId(), AsCharSpan(data.imageURI), uri);
-                    if (CHIP_NO_ERROR != err) {
-                        LogErrorOnFailure(err);
-                        handler->AddStatus(cachedCommandPath, Protocols::InteractionModel::Status::Failure);
+                Commands::QueryImageResponse::Type delegateResponse;
+                ConvertFromQueryImageResponseParms(data, delegateResponse);
+
+                // If update is not available, return the delegate response
+                if (!hasUpdate) {
+                    handler->AddResponse(cachedCommandPath, delegateResponse);
+                    handle.Release();
+                    return;
+                }
+
+                // If there is an update available, return a response with status - Available and set the image Uri in the
+                // response
+                auto fabricIndex = handler->GetSubjectDescriptor().fabricIndex;
+                auto nodeId = handler->GetSubjectDescriptor().subject;
+                CHIP_ERROR err = gOtaSender.PrepareForTransfer(fabricIndex, nodeId);
+                if (CHIP_NO_ERROR != err) {
+
+                    // Handle busy error separately as we have a query image response status that maps to busy
+                    if (err == CHIP_ERROR_BUSY) {
+                        ChipLogError(Controller, "Resource is busy handling another request.");
+                        Commands::QueryImageResponse::Type response;
+                        response.status = static_cast<OTAQueryStatus>(MTROTASoftwareUpdateProviderOTAQueryStatusBusy);
+                        response.delayedActionTime.SetValue(delegateResponse.delayedActionTime.ValueOr(kDelayedActionTimeSeconds));
+                        handler->AddResponse(cachedCommandPath, response);
                         handle.Release();
-                        gOtaSender.ResetState();
                         return;
                     }
-
-                    response.imageURI.SetValue(uri);
-                    handler->AddResponse(cachedCommandPath, response);
+                    LogErrorOnFailure(err);
+                    handler->AddStatus(cachedCommandPath, StatusIB(err).mStatus);
                     handle.Release();
+                    gOtaSender.ResetState();
```

Comment:
Please document why this reset is the right thing to do, even though we have failed to PrepareForTransfer.  Presumably something about how if the state were for a different node we would have taken the early return on CHIP_ERROR_BUSY, so the state is in fact set up for our node if it's set up at all, or something like that?

Reply:
do we need a comment here? in preparefortransfer we set up the timer in configure state and then call the responder preparefortransfer which can fail as well. This is to cover those use cases. I can try to word it in a comment.

Reply:
I think we do need a comment; the logic around ResetState is very finicky and we don't want people to mess it up later.